src/main/java/net/kaczmarzyk/spring/data/jpa/web/WebRequestQueryContext.java

```diff
@@ -51,13 +53,44 @@ public WebRequestQueryContext(NativeWebRequest request) {
 			this.evaluatedJoinFetch = new HashMap<>();
 			request.setAttribute(JOIN_FETCH_ATTRIBUTE_KEY, evaluatedJoinFetch, NativeWebRequest.SCOPE_REQUEST);
 		}
+                
+                this.rootCache = (Map<Pair<String, Root>, javax.persistence.criteria.Join<?, ?>>) request.getAttribute(ROOT_CACHE_ATTRIBUTE_KEY, NativeWebRequest.SCOPE_REQUEST);
+
+		if (this.rootCache == null) {
+			this.rootCache = new HashMap<>();
+			request.setAttribute(ROOT_CACHE_ATTRIBUTE_KEY, rootCache, NativeWebRequest.SCOPE_REQUEST);
+		}
 	}
 
+        @Override
+        public boolean existsJoin(String key, Root<?> root) {
+            Function<Root<?>, Join<?, ?>> value = contextMap.get(key);
+
+            if (value == null) {
+                    Optional<? extends Join<?, ?>> opJoin = root.getJoins().stream().filter(p -> p.getAlias() != null && p.getAlias().equals(key)).findFirst();
+                    if(opJoin.isPresent()){
+                        Join<?, ?> join = opJoin.get();
+                        putLazyVal(key, (r) -> join);
+                        return true;
+                    }
+                    return false;
+            }
+            return true;
```

Comment:
Is it necessary? Why:

```
        @Override
        public boolean existsJoin(String key, Root<?> root) {
            return contextMap.containsKey(key); 
         }
``` 
is not sufficient?

Reply:
Sometimes the join already exists in root but it wasn't added to the context, so I check when it doesn't exist in the context if it doesn't already exist in root, if it exists I just take advantage.